modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -0,0 +1,185 @@
+                                        0);
+    // Add derivative information
+    relative_velocity(0).derivatives() =
+        _secondary_x_dot[_qp].derivatives() - _primary_x_dot[_qp].derivatives();
+    relative_velocity(1).derivatives() =
+        _secondary_y_dot[_qp].derivatives() - _primary_y_dot[_qp].derivatives();
+
+    // Get the component in the tangential direction
+    // TODO: Create a consistent mortar tangential vector field
+    _qp_tangential_velocity = relative_velocity * (_tangents[_qp][0] * _JxW_msm[_qp] * _coord[_qp]);
+  }
+  else
+    _qp_tangential_velocity = std::numeric_limits<Real>::quiet_NaN();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeQpIProperties()
+{
+  // Get the _node_to_weighted_gap map
+  ComputeWeightedGapLMMechanicalContact::computeQpIProperties();
+
+  // Get the _node_to_weighted_tangential_velocity map
+  const auto * const node = _lower_secondary_elem->node_ptr(_i);
+  _node_to_weighted_tangential_velocity[node] += _test[_i][_qp] * _qp_tangential_velocity;
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::residualSetup()
+{
+  // Clear both maps
+  ComputeWeightedGapLMMechanicalContact::residualSetup();
+
+  _node_to_weighted_tangential_velocity.clear();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::jacobianSetup()
+{
+  residualSetup();
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeResidual(const Moose::MortarType mortar_type)
+{
+  if (mortar_type != Moose::MortarType::Lower)
+    return;
+
+  // Fill node to weighted tangential velocity map.
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+  {
+    computeQpProperties();
+    for (_i = 0; _i < _test.size(); ++_i)
+      computeQpIProperties();
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::computeJacobian(const Moose::MortarType mortar_type)
+{
+  // During "computeResidual" and "computeJacobian" we are actually just computing properties on the
+  // mortar segment element mesh. We are *not* actually assembling into the residual/Jacobian. The
+  // property of interest is the map from node to weighted tangential velocity. Computation of the
+  // properties proceeds identically for residual and Jacobian evaluation hence why we simply call
+  // computeResidual here. We will assemble into the residual/Jacobian later from the post() method
+  computeResidual(mortar_type);
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::post()
+{
+  // Enforce weighted gap constraint (calls enforceConstraintOnNode)
+  ComputeWeightedGapLMMechanicalContact::post();
+
+  // Enforce frictional complementarity constraints
+  for (const auto & pr : _node_to_weighted_tangential_velocity)
+  {
+    _weighted_gap_ptr = &_node_to_weighted_gap[pr.first];
+    _tangential_vel_ptr = &pr.second;
+    enforceConstraintOnNode(pr.first);
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::enforceConstraintOnNode(const Node * const node)
+{
+  // Get friction LM index
```

Comment:
So my previous comments stand... don't call the base `post()` from derived `post()` but *do* call the base `enforceConstraintOnNode` from the derived `enforceConstraintOnNode`

Reply:
Yeah, the comments were clear and applied!